File: fatigue_monitor.py

import cv2
import os
import time
import json
import pyttsx3
import base64
import google.generativeai as genai
from PIL import Image
import io

# =================================================================
# GuardDrive AI - Fatigue Monitoring System (PYTHON CORE)
# Author: Senior Frontend/AI Engineer Simulation
# =================================================================

# 1. Configuration
API_KEY = os.environ.get("API_KEY", "YOUR_API_KEY_HERE")
genai.configure(api_key=API_KEY)
model = genai.GenerativeModel('gemini-2.5-flash-lite-latest')

# 2. Hardware Initialization (TTS)
engine = pyttsx3.init()
engine.setProperty('rate', 150)

def speak_alert(message):
    print(f"[HW_TTS] Speaking: {message}")
    engine.say(message)
    engine.runAndWait()

def analyze_frame(frame, lang="CN"):
    """
    Core AI logic: Encodes frame to base64 and calls Gemini Vision API
    """
    _, buffer = cv2.imencode('.jpg', frame)
    img_bytes = buffer.tobytes()
    
    # Construct the instruction
    prompt = f"""
    DRIVER SAFETY MONITOR:
    Analyze this driver face for fatigue. 
    Detect eyes (OPEN/CLOSED), blink frequency, and head posture.
    Return JSON only: {{ "level": "LOW|MODERATE|HEAVY", "reason": "Explain in {'Chinese' if lang=='CN' else 'English'}" }}
    """
    
    try:
        # Simulate Multimodal API Call
        response = model.generate_content([
            {'mime_type': 'image/jpeg', 'data': img_bytes},
            prompt
        ])
        
        # Parse result
        raw_text = response.text.strip()
        if "```json" in raw_text:
            raw_text = raw_text.split("```json")[1].split("```")[0]
            
        return json.loads(raw_text)
    except Exception as e:
        print(f"[AI_ERR] Vision analysis failed: {e}")
        return None

def main():
    print(">>> GUARDDRIVE PYTHON ENGINE STARTING <<<")
    cap = cv2.VideoCapture(0)
    
    last_analysis = 0
    analysis_interval = 5 # seconds
    
    while True:
        ret, frame = cap.read()
        if not ret: break
        
        # Basic HUD
        cv2.putText(frame, "GUARDDRIVE: RUNNING", (20, 40), 
                    cv2.FONT_HERSHEY_SIMPLEX, 0.7, (0, 255, 0), 2)
        cv2.imshow('Driver Monitoring Feed', frame)
        
        current_time = time.time()
        if current_time - last_check > analysis_interval:
            print("[ENGINE] Initiating AI analysis...")
            result = analyze_frame(frame)
            
            if result:
                level = result.get('level', 'LOW')
                reason = result.get('reason', '')
                print(f"[RESULT] Level: {level} | {reason}")
                
                if level == "HEAVY":
                    speak_alert("Warning! Critical fatigue detected. Please pull over.")
                elif level == "MODERATE":
                    speak_alert("Fatigue detected. Suggest taking a rest.")
            
            last_check = current_time
            
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break
            
    cap.release()
    cv2.destroyAllWindows()

if __name__ == "__main__":
    main()

File: fatigue_monitor.cpp

#include <iostream>
#include <string>
#include <vector>
#include <opencv2/opencv.hpp>
#include <windows.h>
#include <sapi.h>

/**
 * GuardDrive AI - C++ Native Engine
 * High-performance fatigue detection with direct hardware hooks.
 */

using namespace cv;
using namespace std;

// Hardware Hook: Windows Text-to-Speech
void HardwareSpeak(const wstring& text) {
    ISpVoice* pVoice = NULL;
    if (FAILED(::CoInitialize(NULL))) return;

    HRESULT hr = ::CoCreateInstance(CLSID_SpVoice, NULL, CLSCTX_ALL, IID_ISpVoice, (void**)&pVoice);
    if (SUCCEEDED(hr)) {
        pVoice->Speak(text.c_str(), 0, NULL);
        pVoice->Release();
    }
    ::CoUninitialize();
}

// Simulated Vision Analysis (In production, use libcurl to call Gemini API)
struct DetectionResult {
    string level;
    string reasoning;
};

DetectionResult NativeAIProcess(Mat& frame) {
    // 1. Pre-process frame (Grayscale, Histogram Equalization)
    Mat gray;
    cvtColor(frame, gray, COLOR_BGR2GRAY);
    equalizeHist(gray, gray);

    // 2. Feature Extraction (Simulated)
    // In full implementation, we convert to Base64 and POST to Google GenAI Endpoint
    cout << "[NATIVE_AI] Image Buffer: " << frame.cols << "x" << frame.rows << " px" << endl;

    // Simulation of API result
    return {"LOW", "No fatigue detected in native AVX2 scan."};
}

int main() {
    cout << "==========================================" << endl;
    cout << "   GUARDDRIVE C++ NATIVE ENGINE V2.0      " << endl;
    cout << "==========================================" << endl;

    VideoCapture cap(0);
    if (!cap.isOpened()) {
        cerr << "FATAL: Primary Camera sensor not found!" << endl;
        return -1;
    }

    Mat frame;
    int cycle = 0;

    while (true) {
        cap >> frame;
        if (frame.empty()) break;

        // HUD overlay
        putText(frame, "C++ ENGINE: ACTIVE", Point(20, 30), FONT_HERSHEY_DUPLEX, 0.6, Scalar(0, 255, 0), 1);
        imshow("GuardDrive Native Feed", frame);

        // Perform AI analysis every 150 frames
        if (cycle % 150 == 0) {
            cout << "[CYCLE] Starting HW-Accelerated Scan..." << endl;
            DetectionResult res = NativeAIProcess(frame);
            cout << "[LOG] " << res.level << ": " << res.reasoning << endl;

            if (res.level == "HEAVY") {
                HardwareSpeak(L"Critical Fatigue Warning. Stopping vehicle.");
            }
        }

        cycle++;
        if (waitKey(30) == 27) break; // ESC to quit
    }

    return 0;
}
